### PR TITLE
Added GetUnitCostTable

### PR DESCRIPTION
**Pull request by 12345swordy**
Last part of #1584

I need someone to review this as the lua table feature is very unintuitive to me.

### CODE REVIEW

**Comment by PicassoCT** [comment]
Needs sassy string as final argument:  You cant afford it honey! https://www.youtube.com/watch?v=8HfYySuUplg

**Comment by 12345swordy** [comment]
I am not interested in your lousy jokes here.

File: rts/Lua/LuaSyncedRead.cpp [hunk 1]
```diff
@@ -4150,6 +4151,27 @@ int LuaSyncedRead::GetUnitCosts(lua_State* L)
...
+ * @number unitID
+ * @treturn nil|{ metal = number, energy = number }, number buildTime
+ * @treturn number buildTime
+ */
+int LuaSyncedRead::GetUnitCostTable(lua_State* L)
+{
+	const CUnit* const unit = ParseInLosUnit(L, __func__, 1);
+	if (unit == nullptr)
+		return 0;
+	lua_createtable(L, 2, 1);
```
Comment (sprunk) [inline rts/Lua/LuaSyncedRead.cpp:4165]: The numbers are the array part (for consecutive numerical keys: 1, 2, 3 ...) and the hash part (all other keys). Here we have two keys, "metal" and "energy", which both go into the hash part:
```suggestion
	lua_createtable(L, 0, 2);
```

File: rts/Lua/LuaSyncedRead.cpp [hunk 2]
```diff
@@ -4150,6 +4151,27 @@ int LuaSyncedRead::GetUnitCosts(lua_State* L)
 	lua_pushnumber(L, unit->cost.energy);
 	return 3;
 }
+/***
+ * @function Spring.GetUnitCostTable
+ * @number unitID
+ * @treturn nil|{ metal = number, energy = number }, number buildTime
```
Comment (sprunk) [inline rts/Lua/LuaSyncedRead.cpp:4157]: `number buildTime` is already listed in the next line.
```suggestion
 * @treturn nil|{ metal = number, energy = number }
```

File: rts/Lua/LuaSyncedRead.cpp [hunk 3]
```diff
@@ -4150,6 +4151,27 @@ int LuaSyncedRead::GetUnitCosts(lua_State* L)
...
+		return 0;
+	lua_createtable(L, 2, 1);
+	lua_pushnumber(L, unit->cost.metal);
+	lua_pushstring(L, "metal");
+	lua_rawset(L, -3);
+	lua_pushnumber(L, unit->cost.energy);
+	lua_pushstring(L, "energy");
+	lua_rawset(L, -3);
+	lua_pushnumber(L, unit->buildTime);
+	return 1;
```
Comment (sprunk) [inline rts/Lua/LuaSyncedRead.cpp:4173]: We return 2 values (the table with regular resources, and buildtime):
```suggestion
	return 2;
```

File: rts/Lua/LuaSyncedRead.cpp [hunk 4]
```diff
@@ -4150,6 +4151,27 @@ int LuaSyncedRead::GetUnitCosts(lua_State* L)
...
+ * @treturn number buildTime
+ */
+int LuaSyncedRead::GetUnitCostTable(lua_State* L)
+{
+	const CUnit* const unit = ParseInLosUnit(L, __func__, 1);
+	if (unit == nullptr)
+		return 0;
+	lua_createtable(L, 2, 1);
+	lua_pushnumber(L, unit->cost.metal);
+	lua_pushstring(L, "metal");
```
Comment (sprunk) [inline rts/Lua/LuaSyncedRead.cpp:4167]: The value should be on top of the stack (i.e. added second), and the key should be the 2nd from the top (i.e. added first)
```suggestion
	lua_pushstring(L, "metal");
	lua_pushnumber(L, unit->cost.metal);
```

File: rts/Lua/LuaSyncedRead.cpp [hunk 5]
```diff
@@ -4150,6 +4151,27 @@ int LuaSyncedRead::GetUnitCosts(lua_State* L)
...
+	lua_rawset(L, -3);
+	lua_pushnumber(L, unit->cost.energy);
+	lua_pushstring(L, "energy");
```
Comment (sprunk) [inline rts/Lua/LuaSyncedRead.cpp:4170]: ditto
```suggestion
	lua_pushstring(L, "energy");
	lua_pushnumber(L, unit->cost.energy);
```

**Review by sprunk — CHANGES_REQUESTED** [review]

File: rts/Lua/LuaSyncedRead.cpp (hunk 1 above)
Comment (12345swordy) [reply rts/Lua/LuaSyncedRead.cpp:4165]: The lua documentation regarding on the lua_createtable function is terrible. Thx.

File: rts/Lua/LuaSyncedRead.cpp (hunk 4 above)
Comment (12345swordy) [reply rts/Lua/LuaSyncedRead.cpp]: Done

File: rts/Lua/LuaSyncedRead.cpp (hunk 5 above)
Comment (12345swordy) [reply rts/Lua/LuaSyncedRead.cpp]: Done

File: rts/Lua/LuaSyncedRead.cpp (hunk 2 above)
Comment (12345swordy) [reply rts/Lua/LuaSyncedRead.cpp]: Done

File: rts/Lua/LuaSyncedRead.cpp (hunk 3 above)
Comment (12345swordy) [reply rts/Lua/LuaSyncedRead.cpp]: Done

**Comment by sprunk** [comment]
Excellent, thanks!

**Pull request merged**